File: examples/enhanced_highlight_demo.py
#!/usr/bin/env python3
"""
Enhanced Highlight Demo with Text Overlays

This demo shows the new enhanced highlight functionality with text overlays.
Run this with Notepad open to see text overlays in different positions.

Requirements: 
- Build the Python bindings: cd bindings/python && pip install -e .
- Or install terminator package if available
"""

import asyncio
import time
import sys
import os

# Add the bindings to path if running from examples folder
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'bindings', 'python'))

try:
    import terminator
except ImportError:
    print("❌ Terminator not found. Please build the Python bindings first:")
    print("   cd bindings/python && pip install -e .")
    sys.exit(1)


async def main():
    print("🎯 Enhanced Highlight Demo - Text Overlays")
    print("=" * 50)
    
    try:
        desktop = terminator.Desktop()
        
        # Step 1: Open Notepad
        print("\n1. Opening Notepad...")
        await desktop.run_command("notepad.exe", "notepad.exe")
        await asyncio.sleep(2)  # Wait for Notepad to open
        
        # Step 2: Find Notepad window and text area
        print("\n2. Finding Notepad elements...")
        
        # Find the Notepad application
        apps = desktop.applications()
        notepad = None
        for app in apps:
            if "notepad" in app.name().lower() or "untitled" in app.name().lower():
                notepad = app
                break
        
        if not notepad:
            print("❌ Notepad not found. Please open Notepad manually and try again.")
            return
        
        print(f"✅ Found Notepad: {notepad.name()}")
        
        # Find the text area (Edit control)
        locator = desktop.locator("role:Edit")
        try:
            text_area = await locator.first()
            print(f"✅ Found text area: {text_area.role()}")
        except terminator.ElementNotFoundError:
            print("❌ Text area not found in Notepad")
            return
        
        # Step 3: Demo different text overlay positions
        print("\n3. Testing different text overlay positions...")
        
        demos = [
            ("TOP", "Top", {"size": 16, "bold": True, "color": 0x000000}),
            ("RIGHT", "Right", {"size": 14, "bold": False, "color": 0x0000FF}),
            ("BOTTOM", "Bottom", {"size": 18, "bold": True, "color": 0x008000}),
            ("LEFT", "Left", {"size": 12, "bold": False, "color": 0xFF0000}),
            ("INSIDE", "Inside", {"size": 20, "bold": True, "color": 0x800080}),
        ]
        
        for text, position, font_style in demos:
            print(f"\n   🔸 Testing {position} position with text '{text}'")
            
            # Highlight with text overlay
            handle = text_area.highlight(
                color=0x00FF00,  # Green border
                duration_ms=3000,  # 3 seconds
                text=text,
                text_position=getattr(terminator.TextPosition, position),
                font_style=terminator.FontStyle(**font_style)
            )
            
            # Wait to see the highlight
            await asyncio.sleep(3.5)
            
            # Demonstrate manual closing
            if position == "Inside":
                print("   📝 Manually closing highlight...")
                handle.close()
                await asyncio.sleep(1)
        
        # Step 4: Demo with context manager (auto-close)
        print("\n4. Testing context manager (auto-close)...")
        
        with text_area.highlight(
            color=0xFF0000,  # Red border
            duration_ms=5000,
            text="AUTO-CLOSE",
            text_position=terminator.TextPosition.TopRight,
            font_style=terminator.FontStyle(size=22, bold=True, color=0xFFFFFF)
        ):
            print("   🔸 Highlight will auto-close when context exits...")
            await asyncio.sleep(2)
            
        print("   ✅ Auto-closed successfully!")
        
        # Step 5: Test different colors and styles
        print("\n5. Testing different font styles and colors...")
        
        style_demos = [
            ("BOLD", {"size": 24, "bold": True, "color": 0x000000}),
            ("BIG", {"size": 32, "bold": False, "color": 0x800000}),
            ("SMALL", {"size": 10, "bold": True, "color": 0x008080}),
        ]
        
        for text, font_style in style_demos:
            print(f"   🔸 Testing font style: {text}")
            
            handle = text_area.highlight(
                color=0x0000FF,
                duration_ms=2000,
                text=text,
                text_position=terminator.TextPosition.Top,
                font_style=terminator.FontStyle(**font_style)
            )
            
            await asyncio.sleep(2.5)
        
        print("\n🎉 Demo completed successfully!")
        print("\nKey features demonstrated:")
        print("  ✅ Text overlays in 9 different positions")
        print("  ✅ Custom font sizes, colors, and bold styling")
        print("  ✅ Manual and automatic highlight closing")
        print("  ✅ Context manager support")
        print("  ✅ White background for text visibility")
        
    except Exception as e:
        print(f"❌ Error during demo: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/enhanced_highlight_demo.ts
#!/usr/bin/env node
/**
 * Enhanced Highlight Demo with Text Overlays
 * 
 * This demo shows the new enhanced highlight functionality with text overlays.
 * Run this with Calculator open to see text overlays in different positions.
 * 
 * Requirements: 
 * - Build the Node.js bindings: cd bindings/nodejs && npm install
 * - Or install terminator package if available
 * - npm install -g tsx (to run TypeScript directly)
 * 
 * Usage: tsx enhanced_highlight_demo.ts
 */

import { dirname, resolve } from 'path';
import { fileURLToPath } from 'url';

// Add the bindings to require path
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const bindingsPath = resolve(__dirname, '..', 'bindings', 'nodejs');

let terminator: any;
try {
    terminator = require(bindingsPath);
} catch (error) {
    console.error('❌ Terminator not found. Please build the Node.js bindings first:');
    console.error('   cd bindings/nodejs && npm install');
    process.exit(1);
}

const { 
    Desktop,
    ElementNotFoundError,
    PlatformError,
    TextPosition,
    FontStyle
} = terminator;

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function main(): Promise<void> {
    console.log('🎯 Enhanced Highlight Demo - Text Overlays');
    console.log('='.repeat(50));
    
    try {
        const desktop = new Desktop();
        
        // Step 1: Open Calculator
        console.log('\n1. Opening Calculator...');
        await desktop.runCommand('calc.exe', 'calc.exe');
        await sleep(2000); // Wait for Calculator to open
        
        // Step 2: Find Calculator window and buttons
        console.log('\n2. Finding Calculator elements...');
        
        // Find the Calculator application
        const apps = desktop.applications();
        let calculator = null;
        for (const app of apps) {
            if (app.name().toLowerCase().includes('calculator')) {
                calculator = app;
                break;
            }
        }
        
        if (!calculator) {
            console.log('❌ Calculator not found. Please open Calculator manually and try again.');
            return;
        }
        
        console.log(`✅ Found Calculator: ${calculator.name()}`);
        
        // Find some number buttons
        const buttonNumbers = ['1', '2', '3', '4', '5'];
        const buttons: Array<{ number: string, element: any }> = [];
        
        for (const num of buttonNumbers) {
            try {
                const locator = desktop.locator(`name:${num}`);
                const button = await locator.first();
                buttons.push({ number: num, element: button });
                console.log(`✅ Found button: ${num}`);
            } catch (e) {
                console.log(`⚠️  Button ${num} not found`);
            }
        }
        
        if (buttons.length === 0) {
            console.log('❌ No calculator buttons found');
            return;
        }
        
        // Step 3: Demo different text overlay positions
        console.log('\n3. Testing different text overlay positions...');
        
        const demos = [
            { text: 'TOP', position: TextPosition.Top, style: { size: 16, bold: true, color: 0x000000 } },
            { text: 'RIGHT', position: TextPosition.Right, style: { size: 14, bold: false, color: 0x0000FF } },
            { text: 'BOTTOM', position: TextPosition.Bottom, style: { size: 18, bold: true, color: 0x008000 } },
            { text: 'LEFT', position: TextPosition.Left, style: { size: 12, bold: false, color: 0xFF0000 } },
            { text: 'INSIDE', position: TextPosition.Inside, style: { size: 20, bold: true, color: 0x800080 } },
        ];
        
        for (let i = 0; i < Math.min(demos.length, buttons.length); i++) {
            const demo = demos[i];
            const button = buttons[i];
            
            console.log(`\n   🔸 Testing ${demo.position} position with text '${demo.text}' on button ${button.number}`);
            
            // Create FontStyle object
            const fontStyle = new FontStyle();
            fontStyle.size = demo.style.size;
            fontStyle.bold = demo.style.bold;
            fontStyle.color = demo.style.color;
            
            // Highlight with text overlay
            const handle = button.element.highlight(
                0x00FF00,  // Green border (color)
                3000,      // 3 seconds (duration_ms)
                demo.text, // text
                demo.position, // text_position
                fontStyle  // font_style
            );
            
            // Wait to see the highlight
            await sleep(3500);
            
            // Demonstrate manual closing for the last one
            if (i === demos.length - 1 || i === buttons.length - 1) {
                console.log('   📝 Manually closing highlight...');
                handle.close();
                await sleep(1000);
            }
        }
        
        // Step 4: Test different colors and styles
        console.log('\n4. Testing different font styles and colors...');
        
        const styledemos = [
            { text: 'BOLD', style: { size: 24, bold: true, color: 0x000000 } },
            { text: 'BIG', style: { size: 32, bold: false, color: 0x800000 } },
            { text: 'SMALL', style: { size: 10, bold: true, color: 0x008080 } },
        ];
        
        for (let i = 0; i < Math.min(styledemos.length, buttons.length); i++) {
            const demo = styledemos[i];
            const button = buttons[i];
            
            console.log(`   🔸 Testing font style: ${demo.text} on button ${button.number}`);
            
            const fontStyle = new FontStyle();
            fontStyle.size = demo.style.size;
            fontStyle.bold = demo.style.bold;
            fontStyle.color = demo.style.color;
            
            const handle = button.element.highlight(
                0x0000FF,      // Blue border
                2000,          // 2 seconds
                demo.text,
                TextPosition.Top,
                fontStyle
            );
            
            await sleep(2500);
        }
        
        // Step 5: Demo longer text with truncation
        console.log('\n5. Testing text truncation...');
        
        if (buttons.length > 0) {
            const fontStyle = new FontStyle();
            fontStyle.size = 14;
            fontStyle.bold = true;
            fontStyle.color = 0x000000;
            
            const handle = buttons[0].element.highlight(
                0xFF00FF,      // Magenta border
                4000,          // 4 seconds
                'This text will be truncated to 10 chars', // Long text
                TextPosition.Bottom,
                fontStyle
            );
            
            console.log('   🔸 Testing text truncation (should show "This te...")');
            await sleep(4500);
        }
        
        console.log('\n🎉 Demo completed successfully!');
        console.log('\nKey features demonstrated:');
        console.log('  ✅ Text overlays in different positions');
        console.log('  ✅ Custom font sizes, colors, and bold styling');
        console.log('  ✅ Manual highlight closing');
        console.log('  ✅ Text truncation to 10 characters');
        console.log('  ✅ White background for text visibility');
        
    } catch (error) {
        console.error('❌ Error during demo:', error);
        if (error instanceof Error) {
            console.error(error.stack);
        }
    }
}

// Handle unhandled promise rejections
process.on('unhandledRejection', (reason, promise) => {
    console.error('Unhandled Rejection at:', promise, 'reason:', reason);
});

main().catch(err => {
    console.error('Fatal error:', err);
    process.exit(1);
});